File: stgit/commands/pull.py
from stgit.argparse import opt
from stgit.commands.common import (
    CmdException,
    DirectoryGotoTopLevel,
    git_commit,
    post_rebase,
    prepare_rebase,
    rebase,
)
from stgit.config import GitConfigException, config
from stgit.lib.git import MergeConflictException, RepositoryException
from stgit.out import out

__copyright__ = """
Copyright (C) 2005, Catalin Marinas <catalin.marinas@gmail.com>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License version 2 as
published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, see http://www.gnu.org/licenses/.
"""

help = 'Pull changes from a remote repository'
kind = 'stack'
usage = ['[options] [--] [<repository>]']
description = """
Pull the latest changes from the given remote repository (defaulting
to branch.<name>.remote, or 'origin' if not set). This command works
by popping all the patches from the stack, pulling the changes in the
parent repository, setting the base of the stack to the latest parent
HEAD and pushing the patches back (unless '--nopush' is specified).
The 'push' operation can fail if there are conflicts. They need to be
resolved and the patch pushed again.

Check the 'git fetch' documentation for the <repository> format."""

args = ['repo']
options = [
    opt(
        '-n',
        '--nopush',
        action='store_true',
        short='Do not push the patches back after pulling',
    ),
    opt(
        '-m',
        '--merged',
        action='store_true',
        short='Check for patches merged upstream',
    ),
]

directory = DirectoryGotoTopLevel()


def pull(repository, remote_repository):
    """Pull changes from remote repository using 'git pull' by default."""
    command = (
        config.get('branch.%s.stgit.pullcmd' % repository.current_branch_name)
        or config.get('stgit.pullcmd')
    ).split()
    command.append(remote_repository)
    runner = repository.run(command)

    # Assumption: pullcmd will return 1 for a conflict and some non-1 code for
    # other errors. This is probably an incorrect assumption because pullcmd
    # may not be `git pull` and even `git pull` does not make any clear
    # guarantees about its return code. However, the consequence of any of this
    # going wrong will be that `stg pull` will return 3 (conflict) when 2
    # (generic error) might be more correct.
    runner.returns([0, 1]).run()
    if runner.exitcode:
        raise MergeConflictException([])

    repository.refs.reset_cache()


def fetch(repository, remote_repository):
    """Fetch changes from remote repository using 'git fetch' by default"""
    command = (
        config.get('branch.%s.stgit.fetchcmd' % repository.current_branch_name)
        or config.get('stgit.fetchcmd')
    ).split()
    command.append(remote_repository)
    repository.run(command).run()


def func(parser, options, args):
    """Pull the changes from a remote repository
    """
    repository = directory.repository
    iw = repository.default_iw
    stack = repository.get_stack()
    policy = config.get('branch.%s.stgit.pull-policy' % stack.name) or config.get(
        'stgit.pull-policy'
    )

    if policy == 'rebase':
        # parent is local
        if len(args) == 1:
            parser.error(
                'specifying a repository is meaningless for policy="%s"' % (policy,)
            )
        elif len(args) > 0:
            parser.error('incorrect number of arguments')
    else:
        # parent is remote
        if len(args) > 1:
            parser.error('incorrect number of arguments')

        if len(args) >= 1:
            remote_name = args[0]
        else:
            remote_name = stack.parent_remote

    if stack.protected:
        raise CmdException('This branch is protected. Pulls are not permitted')

    if policy not in ['pull', 'fetch-rebase', 'rebase']:
        raise GitConfigException('Unsupported pull-policy "%s"' % policy)

    applied = stack.patchorder.applied

    retval = prepare_rebase(stack, 'pull')
    if retval:
        return retval

    # pull the remote changes
    if policy == 'pull':
        out.info('Pulling from "%s"' % remote_name)
        pull(repository, remote_name)
    elif policy == 'fetch-rebase':
        out.info('Fetching from "%s"' % remote_name)
        fetch(repository, remote_name)
        try:
            target = repository.rev_parse('FETCH_HEAD')
        except RepositoryException:
            out.error(
                'Could not find the remote head to rebase onto - '
                'fix branch.%s.merge in .git/config' % stack.name
            )
            out.error('Pushing any patches back...')
            post_rebase(stack, applied, 'pull', check_merged=False)
            raise

        rebase(stack, iw, target)
    elif policy == 'rebase':
        value = config.get('branch.%s.stgit.parentbranch' % stack.name)
        if value:
            parent_commit = git_commit(value, repository)
        else:
            try:
                parent_commit = repository.rev_parse('heads/origin')
            except RepositoryException:
                raise CmdException('Cannot find a parent branch for "%s"' % stack.name)
            else:
                out.warn(
                    'No parent branch declared for stack "%s", defaulting to'
                    '"heads/origin".' % stack.name,
                    'Consider setting "branch.%s.stgit.parentbranch" with '
                    '"git config".' % stack.name,
                )
        rebase(stack, iw, parent_commit)

    if not options.nopush:
        post_rebase(stack, applied, 'pull', check_merged=options.merged)

    # maybe tidy up
    if config.getbool('stgit.keepoptimized'):
        repository.repack()

File: src/cmd/pull.rs
// SPDX-License-Identifier: GPL-2.0-only

//! `stg pull` implementation.

use std::{fmt::Display, str::FromStr};

use anyhow::{anyhow, Context, Result};
use bstr::ByteSlice;
use clap::{Arg, ArgMatches};

use crate::{
    argset,
    color::get_color_stdout,
    ext::RepositoryExtended,
    print_info_message,
    stack::{InitializationPolicy, Stack, StackAccess, StackStateAccess},
    stupid::Stupid,
};

pub(super) const STGIT_COMMAND: super::StGitCommand = super::StGitCommand {
    name: "pull",
    category: super::CommandCategory::StackManipulation,
    make,
    run,
};

fn make() -> clap::Command {
    clap::Command::new(STGIT_COMMAND.name)
        .about("Pull changes from a remote repository")
        .long_about(
            "Pull the latest changes from a remote repository.\n\
             \n\
             The remote repository may be specified on the command line, but defaults \
             to branch.<name>.remote from the git configuration, or \"origin\" if not \
             configured.\n\
             \n\
             This command works by popping all currently applied patches from the \
             stack, pulling the changes from the remote repository, updating the stack \
             base to the new remote HEAD, and finally pushing all formerly applied \
             patches back onto the stack. Merge conflicts may occur during the final \
             push step. Those conflicts need to be resolved manually.\n\
             \n\
             See git-fetch(1) for the format of remote repository argument.
             ",
        )
        .arg(Arg::new("repository").help("Repository to pull from"))
        .arg(
            Arg::new("nopush")
                .long("nopush")
                .short('n')
                .help("Do not push back patches after pulling")
                .action(clap::ArgAction::SetTrue)
                .conflicts_with("merged"),
        )
        .arg(argset::merged_arg().long_help(
            "Check for patches that may have been merged upstream.\n\
             \n\
             When pushing-back patches, each patch is checked to see if its changes \
             already exist in the just-pulled upstream changes. If a patch's changes \
             have already been merged upstream, the patch will still exist in the \
             stack, but become empty after the pull operation.",
        ))
        .arg(argset::push_conflicts_arg())
}

enum PullPolicy {
    Pull,
    Rebase,
    FetchRebase,
}

impl FromStr for PullPolicy {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "pull" => Ok(PullPolicy::Pull),
            "rebase" => Ok(PullPolicy::Rebase),
            "fetch-rebase" => Ok(PullPolicy::FetchRebase),
            _ => Err(anyhow!("unsupported pull-policy `{s}`")),
        }
    }
}

impl Display for PullPolicy {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            PullPolicy::Pull => "pull",
            PullPolicy::Rebase => "rebase",
            PullPolicy::FetchRebase => "fetch-rebase",
        }
        .fmt(f)
    }
}

fn run(matches: &ArgMatches) -> Result<()> {
    let repo = gix::Repository::open()?;
    let stack = Stack::current(&repo, InitializationPolicy::RequireInitialized)?;
    let stupid = repo.stupid();
    let branch_name = stack.get_branch_name().to_string();
    let config = repo.config_snapshot();
    let policy = PullPolicy::from_str(
        &config
            .string_by(
                "branch",
                Some(format!("{branch_name}.stgit").as_str().into()),
                "pull-policy",
            )
            .or_else(|| config.string("stgit.pull-policy"))
            .map(|bs| bs.to_str_lossy().to_string())
            .unwrap_or_else(|| "pull".to_string()),
    )?;

    let allow_push_conflicts = argset::resolve_allow_push_conflicts(&config, matches);

    let parent_remote;
    let remote_name = match policy {
        PullPolicy::Rebase => {
            if matches.get_one::<String>("repository").is_some() {
                return Err(anyhow!(
                    "specifying a repository is meaningless for `{policy}` pull-policy"
                ));
            }
            None
        }
        PullPolicy::Pull | PullPolicy::FetchRebase => {
            parent_remote = config
                .string_by("branch", Some(branch_name.as_str().into()), "remote")
                .and_then(|bs| bs.to_str().map(str::to_string).ok());
            let remote_name = matches
                .get_one::<String>("repository")
                .cloned()
                .or_else(|| parent_remote.clone());
            if remote_name.is_none() {
                return Err(anyhow!(
                    "no tracking information for the current branch;\n\
                     please specify the remote repository to pull from."
                ));
            }
            remote_name
        }
    };

    if stack.is_protected(&config) {
        return Err(anyhow!("this branch is protected; pulls are not permitted"));
    }

    stupid.statuses(None)?.check_index_and_worktree_clean()?;
    stack.check_head_top_mismatch()?;

    let applied = stack.applied().to_vec();

    stack
        .setup_transaction()
        .use_index_and_worktree(true)
        .with_output_stream(get_color_stdout(matches))
        .transact(|trans| {
            trans.pop_patches(|pn| applied.contains(pn))?;
            Ok(())
        })
        .execute("pull (pop)")?;

    let rebase_target = match policy {
        PullPolicy::Pull => {
            let pull_cmd = config
                .string_by(
                    "branch",
                    Some(format!("{branch_name}.stgit").as_str().into()),
                    "pullcmd",
                )
                .or_else(|| config.string("stgit.pullcmd"))
                .and_then(|bs| bs.to_str().map(str::to_string).ok())
                .unwrap_or_else(|| "git pull".to_string());
            let remote_name = remote_name.unwrap();
            print_info_message(matches, &format!("Pulling from `{remote_name}`"));
            if !stupid.user_pull(&pull_cmd, &remote_name)? {
                return Err(super::Error::CausedConflicts(
                    "pull resulted in conflicts".to_string(),
                )
                .into());
            }
            None
        }
        PullPolicy::FetchRebase => {
            let fetch_cmd = config
                .string_by(
                    "branch",
                    Some(format!("{branch_name}.stgit").as_str().into()),
                    "fetchcmd",
                )
                .or_else(|| config.string("stgit.fetchcmd"))
                .and_then(|bs| bs.to_str().map(str::to_string).ok())
                .unwrap_or_else(|| "git fetch".to_string());
            let remote_name = remote_name.unwrap();
            print_info_message(matches, &format!("Fetching from `{remote_name}`"));
            stupid.user_fetch(&fetch_cmd, &remote_name)?;
            let target_id = repo
                .find_reference("FETCH_HEAD")
                .context("finding `FETCH_HEAD`")?
                .peel_to_commit()
                .context("peeling `FETCH_HEAD` to commit")?
                .id;
            Some(target_id)
        }
        PullPolicy::Rebase => {
            let parent_branch_name = config.string_by(
                "branch",
                Some(format!("{branch_name}.stgit").as_str().into()),
                "parentbranch",
            );
            let parent_branch_name = parent_branch_name.as_ref().and_then(|bs| bs.to_str().ok());

            let parent_object = if let Some(name) = parent_branch_name {
                repo.rev_parse_single_ex(name)?.object()?
            } else {
                repo.rev_parse_single("heads/origin")
                    .map_err(|_| anyhow!("cannot find a parent branch for `{branch_name}`"))?
                    .object()?
            };
            let parent_commit = parent_object
                .peel_tags_to_end()
                .context("peel parent object to commit")?
                .try_into_commit()?;
            Some(parent_commit.id)
        }
    };

    if let Some(rebase_target) = rebase_target {
        let rebase_cmd = config
            .string_by(
                "branch",
                Some(format!("{branch_name}.stgit").as_str().into()),
                "rebasecmd",
            )
            .or_else(|| config.string("stgit.rebasecmd"))
            .and_then(|bs| bs.to_str().map(str::to_string).ok())
            .unwrap_or_else(|| "git reset --hard".to_string());
        print_info_message(matches, &format!("Rebasing to `{rebase_target}`"));
        stupid.user_rebase(&rebase_cmd, rebase_target)?;
    }

    // The above pull and rebase action may have moved the stack's branch reference,
    // so we initialize the stack afresh.
    let stack = Stack::current(&repo, InitializationPolicy::RequireInitialized)?;
    let stack = if stack.is_head_top() {
        stack
    } else {
        // Record a new stack state with updated head since the pull moved the head.
        stack.log_external_mods(Some("pull"))?
    };

    if !matches.get_flag("nopush") {
        stack.check_head_top_mismatch()?;
        let check_merged = matches.get_flag("merged");
        stack
            .setup_transaction()
            .use_index_and_worktree(true)
            .allow_push_conflicts(allow_push_conflicts)
            .with_output_stream(get_color_stdout(matches))
            .transact(|trans| trans.push_patches(&applied, check_merged))
            .execute("pull (reapply)")?;
    }

    if config.boolean("stgit.keepoptimized").unwrap_or(false) {
        stupid.repack()?;
    }

    Ok(())
}
